feat: Support Sidebar - Max

slshults opened the pull request:
## Problem

Support AI Chat for the support sidebar. Help users help themselves, reduce support ticket volume.

![CleanShot 2024-12-19 at 18 27 46@2x](https://github.com/user-attachments/assets/1be2d4ee-5290-4982-bc7d-418ddad36862) 
![CleanShot 2024-12-19 at 18 28 13@2x](https://github.com/user-attachments/assets/5403b762-ed3b-4712-bd2c-3197a51ebd0d)

![CleanShot 2024-12-19 at 18 31 12@2x](https://github.com/user-attachments/assets/a7cdab64-65bf-4d2d-9dda-014a85314c32)
![CleanShot 2024-12-19 at 18 31 32@2x](https://github.com/user-attachments/assets/3cf0d997-37c4-4633-be62-3e87c95c2c99)

## Does this work well for both Cloud and self-hosted?

Cloud only

## How did you test this code?
Lots of chatting with Max, in two browsers concurrently.

Feature flag in place for a bit more internal testing before we give access to a small set of customers initially.


Edit: Once this review is done, I'll be submitting another PR for [this more recent commit](https://github.com/PostHog/posthog/commit/ab99ce45bd97331dd9c546fe5af92fde701d9bf7)

slshults committed:
Support sidebar Max AI - not yet integrated

slshults committed:
Fixed the `Send` button, no longer looks disabled.

slshults committed:
Fixed autoscrolling

slshults committed:
Created an exception to allow  to work on links in Max's responses. Also added 1 line to URL validation in the system prompt.

slshults committed:
Put a stop to the weird autoscroll upward and
flashy/blinky behavior seen while typing
after Max had included a codeblock in a response.

slshults committed:
Restoring error handling for 500 errors

slshults committed:
On the support sidebar, hid the docs ToC behind
a collapsible, "Explore the docs", because
less scrolling is more betterer.

slshults committed:
Fixed links to the support form, generated by Max, so that they open in the sidepanel instead of in a new tab.

slshults committed:
Starting to integrate Max. He works in the previous commit, broken in this commit. Will be pairing with others to revive him.

slshults committed:
A step in bringing this branch closer to up-to-date with master

slshults committed:
Another step in bringing  up to date with

slshults committed:
another step in bringing the  branch up to date with master. Server can be run on the local with this commit, but can't run Max yet, need to know how to handle secrets correctly.

slshults committed:
Max is now integrated and working on my local, steps remain (secrets etc.)

slshults committed:
catching up with master

Twixes committed:
Simplify settings

slshults committed:
merged Max's requirements, fixed session handling, simplified conversation management, improved token and cache tracking.

slshults committed:
Merge remote-tracking branch 'origin/master' into support-sidebar-max-integration

slshults committed:
Raising the feature flag. ⛳️

slshults committed:
catching up merged requirements-dev.txt

Twixes committed:
Merge branch 'master' into support-sidebar-max-integration

Twixes committed:
Remove redundant new deps

Twixes committed:
Roll back some redundant frontend changes

Twixes committed:
Roll back some redundant frontend changes

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update UI snapshots for `chromium` (1)

Twixes committed:
Merge branch 'support-sidebar-max-integration' of https://github.com/PostHog/posthog into support-sidebar-max-integration

slshults committed:
steps toward moving functions from python to django. The connection to anthropic API doesn't work in this commit, pushing it to pair with Michael on it.

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update UI snapshots for `chromium` (1)

Twixes committed:
Merge branch 'master' into support-sidebar-max-integration

Twixes committed:
Remove some dead code

Twixes committed:
Fix views.py

slshults committed:
Merge branch 'master' into support-sidebar-max-integration

slshults committed:
Finished moving functions from Flask to Django, fixed resulting bugs, removed dead code, included 'title' in things hidden by the feature flag. (still need help from infra with encrypting Max's secrets.)

slshults committed:
Merge remote-tracking branch 'origin/master' into support-sidebar-max-integration

slshults committed:
Resolving 'ERR_PNPM_LOCKFILE_CONFIG_MISMATCH'

github-actions[bot] committed:
Update query snapshots

slshults commented on frontend/src/lib/lemon-ui/Link/Link.tsx:
```diff
@@ -21,6 +21,8 @@ export type LinkProps = Pick<React.HTMLProps<HTMLAnchorElement>, 'target' | 'cla
     to?: string | [string, RoutePart?, RoutePart?]
     /** If true, in-app navigation will not be used and the link will navigate with a page load */
     disableClientSideRouting?: boolean
+    /** If true, docs links will not be opened in the docs panel */
+    disableDocsPanel?: boolean
```
This is part of an exception to allow links generated by Max to open in a new tab, since there's code to force links from the `Docs` panel to ignore `target="_blank`.  This is part of the exception carved out for links that Max includes in his answers so they'll open in a new tab (currently we'd lose the chat session if the links opened in the sidepanel instead of a new tab.)

The key difference:
- `disableDocsPanel` is specific to documentation links and controls whether they open in the side panel
- `disableClientSideRouting` is a broader setting that affects all navigation, determining whether links should trigger a full page load or use client-side routing

slshults commented on .gitignore:
```diff
@@ -68,6 +68,13 @@ plugin-transpiler/dist
 *.log
 # pyright config (keep this until we have a standardized one)
 pyrightconfig.json
+
+# Max-specific entries
+ee/support_sidebar_max/max-venv/
+ee/support_sidebar_max/.vscode
+ee/support_sidebar_max/.vscode/settings.json
+max-test-venv/
+ee/support_sidebar_max/.env
```
Some functions were moved, but the main python script is still in use and still in `ee/support_sidebar_max` (`sidebar_max_AI.py`), so I still don't want to be pushing those items from my local. (I can and will remove `max-test-venv` though, done with that.)

slshults commented on package.json:
```diff
@@ -66,6 +66,7 @@
         "mypy-baseline-sync": "mypy -p posthog | mypy-baseline sync"
     },
     "dependencies": {
+        "@anthropic-ai/sdk": "^0.33.1",
```
Good catch, thanks! The Node.js/JavaScript version of the SDK, while in our Python code we're using the Python SDK. Shall remove.

slshults committed:
chore: update dependencies from master

slshults committed:
Merge branch 'support-sidebar-max-integration' of https://github.com/PostHog/posthog into support-sidebar-max-integration

slshults committed:
Merge branch 'master' into support-sidebar-max-integration

slshults committed:
Reducing token usage by 33% to 66% per turn with a clever system prompt change

slshults committed:
upgrading to Anthropic Python SDK v 0.42.0, corrected some logging, dialed back logging verbosity.

slshults committed:
Merge branch 'master' into support-sidebar-max-integration

skoob13 reviewed (changes requested):
Looks great! Everything works locally. The agent's persona is perfect🥇 

There are two things we need to address before shipping it:
- pnpm lockfiles for `frontend/@posthog/apps-common/package.json` and `plugin-server/package.json`.
- Storage of the conversations.

Blocking calls can be fixed later :)

slshults commented:
> Looks great! Everything works locally. The agent's persona is perfect🥇
> 
> There are two things we need to address before shipping it:
> 
> * pnpm lockfiles for `frontend/@posthog/apps-common/package.json` and `plugin-server/package.json`.
> * Storage of the conversations.
> 
> Blocking calls can be fixed later :)

Thanks @skoob13 !   

I think I fixed the `package.json` issues, and reverted the erroneous `pnpm-lock.yaml` change in a subsequent commit, (will double-check before moving the PR to a later commit.)

My thinking had been to add storage after we get the MVP up and running. The Anthropic API has 200K+ token context window, plus we're using prompt-caching for the system prompt, search tool, and users initial prompt, so you can have a good 5 to 10 turn conversation without losing context.  Anthropic's prompt caching is stored on their servers (flushed every five mins for each session.) 

I'm thinking this would also mean we could get the MVP up before addressing the load balancing challenges, but let me know if I'm confused on that. (If I'm confused and we do need to be storing conversations for MVP, then yes storing the conversation history on frontend, so the view will be stateless, seems like the way to go. I may have questions. 😊)

With those things in mind, does it seem viable to you wait to add conversation storage after MVP, or is load balancing going to break conversations without it?

Thanks for the `sync_to_async` link! Yes, four workers will not be enough.

skoob13 commented:
> > Looks great! Everything works locally. The agent's persona is perfect🥇
> > There are two things we need to address before shipping it:
> > 
> > * pnpm lockfiles for `frontend/@posthog/apps-common/package.json` and `plugin-server/package.json`.
> > * Storage of the conversations.
> > 
> > Blocking calls can be fixed later :)
> 
> Thanks @skoob13 !
> 
> I think I fixed the `package.json` issues, and reverted the erroneous `pnpm-lock.yaml` change in a subsequent commit, (will double-check before moving the PR to a later commit.)
> 
> My thinking had been to add storage after we get the MVP up and running. The Anthropic API has 200K+ token context window, plus we're using prompt-caching for the system prompt, search tool, and users initial prompt, so you can have a good 5 to 10 turn conversation without losing context. Anthropic's prompt caching is stored on their servers (flushed every five mins for each session.)
> 
> I'm thinking this would also mean we could get the MVP up before addressing the load balancing challenges, but let me know if I'm confused on that. (If I'm confused and we do need to be storing conversations for MVP, then yes storing the conversation history on frontend, so the view will be stateless, seems like the way to go. I may have questions. 😊)
> 
> With those things in mind, does it seem viable to you wait to add conversation storage after MVP, or is load balancing going to break conversations without it?
> 
> Thanks for the `sync_to_async` link! Yes, four workers will not be enough.

The main problem now is that the viewset assumes that it only runs on a single machine (stateful). When we run this code in production, it will run across many EC2 instances, so user requests won't always be routed to the same machine. This effectively means that every request to the endpoint after the first one can fail because it might reach the same machine with the saved conversation state or not.

slshults committed:
Merge master into support-sidebar-max-integration

slshults committed:
refactor: move conversation storage to Redis cache

slshults committed:
Merge remote-tracking branch 'origin/master' into support-sidebar-max-integration

github-actions[bot] committed:
Update UI snapshots for `chromium` (1)

github-actions[bot] committed:
Update UI snapshots for `chromium` (1)

slshults commented:
@skoob13 Conversation storage added in [280fe34](https://github.com/PostHog/posthog/commit/280fe3436bdd20500fb5ef9f39b015dd2051c1d9) . Works on my local, not sure how to test it properly, suggestions welcome.

slshults committed:
feat(support-sidebar-max): Add Django sync-to-async support to MaxChatViewSet

slshults committed:
Merge branch 'master' into support-sidebar-max-integration

slshults committed:
Merge branch 'support-sidebar-max-integration' of https://github.com/PostHog/posthog into support-sidebar-max-integration

skoob13 commented:
Let's fix the CI and comments above, and it's good to go. Nice work with caching!

slshults committed:
Merge master into support-sidebar-max-integration

slshults committed:
refactor: rename files and improve error handling - rename system_prompt/AI files to match conventions, update imports, improve error security

slshults committed:
refactor: rename files and improve error handling - rename system_prompt/AI files to match

slshults committed:
Merge branch 'master' into support-sidebar-max-integration

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update UI snapshots for `chromium` (1)

slshults committed:
chore: reset pnpm lock files to match master

slshults committed:
Merge branch 'support-sidebar-max-integration' of https://github.com/PostHog/posthog into support-sidebar-max-integration

slshults commented:
Thanks @skoob13 ! Changes made (checks are still running, but I'll revisit if any fail.)

skoob13 reviewed (changes requested):
Almost there. Three things:
- requirements.in/txt got lost. We don't have beautifulsoup and Anthropic SDK installed.
- You need to run Ruff locally to pass the CI and commit the format changes: `ruff check --fix`.
- Some tests seem to be broken, but it might just be a flaky CI.

and low-priority nits if you have time.

skoob13 commented on ee/urls.py:
```diff
@@ -1,16 +1,16 @@
+from django.views.decorators.csrf import csrf_exempt
 from ee.api import integration
+from ee.support_sidebar_max.views import MaxChatViewSet
+from .api.rbac import organization_resource_access, role
```
Some endpoints were removed, so that's why CI fails.

slshults committed:
Merge branch 'master' into support-sidebar-max-integration

slshults committed:
restored beautifulsoup and Anthropic SDK to requirements.txt, renamed prompt file

slshults commented:
Thanks much @skoob13 , each of those is addressed a commit I just pushed, keeping an eye on the checks...

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update UI snapshots for `chromium` (1)

github-actions[bot] committed:
Update UI snapshots for `chromium` (1)

skoob13 committed:
fix: missing route and formatting

skoob13 reviewed (approved):
Good to go. Great job!

slshults committed:
Merge branch 'master' into support-sidebar-max-integration

slshults committed:
fixes 1 and 2 of 6 static type errors

slshults committed:
fixes the 3rd of 6 static type errors

slshults committed:
Fixes the last of six static type issues

slshults committed:
Merge branch 'master' into support-sidebar-max-integration
